### PR TITLE
prov/verbs: refactoring of rdm buffers

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
This series of patches is a deep refactoring of addressing and memory layout of preregistered buffers. It reduces HW requirement for wr.imm_data.

@a-ilango @shefty

### CODE REVIEW

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 1]
```diff
@@ -307,23 +316,16 @@ struct fi_ibv_rdm_tagged_postponed_entry {
...
-enum fi_ibv_rdm_buffer_status {
-	BUF_STATUS_FREE = 0,
-	BUF_STATUS_BUSY
-};
-
+/*
+ * Size of the structure is performance sensitive. TODO: resize to 4 bytes.
+ */
 struct fi_ibv_rdm_buffer_service_data {
 	volatile enum fi_ibv_rdm_buffer_status status;
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: This is not part of this patch, but I noticed this as part of the review.  The use of volatile is almost always wrong unless it is a memory address that an IO device can update.

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 2]
```diff
@@ -307,23 +316,16 @@ struct fi_ibv_rdm_tagged_postponed_entry {
...
 
-#define FI_IBV_RDM_BUFF_SERVICE_DATA_SIZE				\
-	(sizeof(struct fi_ibv_rdm_buffer_service_data)			\
-		< FI_IBV_RDM_MEM_ALIGNMENT ? FI_IBV_RDM_MEM_ALIGNMENT	\
-		: (sizeof(struct fi_ibv_rdm_buffer_service_data) +	\
-		  (FI_IBV_RDM_MEM_ALIGNMENT -				\
-		   sizeof(struct fi_ibv_rdm_buffer_service_data) %	\
-		   FI_IBV_RDM_MEM_ALIGNMENT)))
+#define FI_IBV_RDM_BUFF_SERVICE_DATA_SIZE	\
+	(sizeof(struct fi_ibv_rdm_buffer_service_data))
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Having a define replace sizeof unnecessarily complicates the code. Just use sizeof directly.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c [hunk 3]
```diff
@@ -642,52 +626,58 @@ void check_and_repost_receives(struct fi_ibv_rdm_ep *ep,
...
+				}
+
+				check_and_repost_receives(ep, conn);
+
+				if (fi_ibv_rdm_get_buffer_status((char*) rbuf) == BUF_STATUS_RECVED &&
+				    fi_ibv_rdm_buffer_check_pkt_len((char*) rbuf, wc[i].byte_len))
+				{
+					do {
+						fi_ibv_rdm_tagged_process_recv(ep, conn, fi_ibv_rdm_get_buff_service_data((char*) rbuf)->pkt_len, rbuf);
+						fi_ibv_rdm_set_buffer_status((char*) rbuf, BUF_STATUS_FREE);
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: This is nested too deep.  Consider moving inner loops into separate functions for readability.  (The compiler will fold everything back in.)

File: prov/verbs/src/ep_rdm/verbs_utils.h [hunk 4]
```diff
@@ -64,14 +64,12 @@ struct fi_ibv_msg_ep;
...
 #define FI_IBV_RDM_TAGGED_DFLT_BUFFER_NUM (8)
 
 #define FI_IBV_RDM_DFLT_BUFFER_SIZE					\
-	((8 * 1024 + FI_IBV_RDM_BUFF_SERVICE_DATA_SIZE) +		\
-	 (8 * 1024 + FI_IBV_RDM_BUFF_SERVICE_DATA_SIZE) %		\
-	  FI_IBV_RDM_BUF_ALIGNMENT)
+	(3 * FI_IBV_RDM_BUF_ALIGNMENT)
 
 #define FI_IBV_RDM_DFLT_BUFFERED_SSIZE					\
 	(FI_IBV_RDM_DFLT_BUFFER_SIZE -					\
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_utils.h]: This is a general comment about this commit.  There are a lot of white space changes that add noise to the overall patch, and the original formatting of the code is better aligned with the coding guidelines and style.

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c [hunk 5]
```diff
@@ -145,6 +145,12 @@ fi_ibv_rdm_tagged_prepare_conn_memory(struct fi_ibv_rdm_ep *ep,
 				(void **) &conn->rbuf_mem_reg, size);
 	assert(conn->r_mr);
 
+	conn->ack_mr = ibv_reg_mr(ep->domain->pd, &conn->sbuf_ack_status,
+		sizeof(conn->sbuf_ack_status),
+		IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE);
+
+	assert(conn->ack_mr);
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: This should have proper error handling, not just debug assertions.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c [hunk 6]
```diff
@@ -718,46 +718,54 @@ static inline int fi_ibv_rdm_tagged_poll_recv(struct fi_ibv_rdm_ep *ep)
...
-							.ep = ep
-						};
-
-						fi_ibv_rdm_tagged_req_hndl
-							(request, FI_IBV_EVENT_SEND_GOT_LC, &data);
-					}
-				} else {
-					goto wc_error;
-				}
+				err &= fi_ibv_rdm_process_send_wc(ep, &wc[i]);
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: Why the logical and, versus just setting err directly?  We break from the loop if err is non-zero.

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 7]
```diff
@@ -110,10 +110,10 @@ do {                                                                	\
 	 PEND_SEND_IS_LIMITED(_ep))
 
 struct fi_ibv_rdm_header {
-	uint64_t imm_data;          // TODO: not implemented
+/*	uint64_t imm_data; TODO: not implemented */
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Is this changing the wire protocol?

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 8]
```diff
@@ -196,7 +196,22 @@ fi_ibv_rdm_tagged_zero_request(struct fi_ibv_rdm_tagged_request *request)
 void fi_ibv_rdm_tagged_print_request(char *buf,
 				     struct fi_ibv_rdm_tagged_request *request);
 
+typedef short fi_ibv_rdm_buf_status_t;
+#define BUF_STATUS_FREE		((fi_ibv_rdm_buf_status_t) 0)
+#define BUF_STATUS_BUSY		((fi_ibv_rdm_buf_status_t) 1)
+#define BUF_STATUS_RECVED	((fi_ibv_rdm_buf_status_t) 2)
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: typedefs should be avoided unless there's an architectural reason for them.  Weren't these enums in earlier patches?

**Comment by shefty** [comment]
I had some comments based on reviewing individual commits, some of which won't apply to the final set of changes.  I have some concerns, but I'm okay if you want to address those in subsequent patches, since these changes are contained to the verbs RDM support only.  I'll let @a-ilango provide more in depth review of the changes, as I'm not overly familiar with the RDM code base.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Buffer's status is a part of buffers control. Remote side is updating this address by RDMA_WRITE operation to release then it's can be reused.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 2 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: This code is replaced with "offsetof" because of sensitivity to alignment.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c (hunk 3 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: It's done in one from the next patches. This was a temporary code during refactoring.

File: prov/verbs/src/ep_rdm/verbs_utils.h (hunk 4 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_utils.h]: These lines were changed anyway because of type changing or something else. That is more preferable alignment of args, to opening bracket by spaces or common tabulation?

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c (hunk 5 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: Thanks, good point! I'll fix it.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c (hunk 6 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: I decided to move all extra logic to error handling for both polling (send and recv) because this is a critical path. Currently, there are no matter which exact WC returned by ibv_poll_cq is erroneous. We can't restore QP state yet. So, we just process all what we can when go to error handling code block to print error message for correspond WC and return error code.
In addition, buffer's service data and pkt header on a sender side contains all required data for re transmitting after restoring of QP state. It's planned to be implemented.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 7 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: No, this field was reserved for implementing of OFI tagged imm_data. It's not supported yet.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 8 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: enum has 32-bit size by default, I replaced it by short and corresponding definitions from performance reason. Before this series, buffer status has generally local nature (excepting releasing). Now, it's a part of protocol which is sending every time with header and users data. So, I would say, service data of a buffer (including status) is a things of lower level protocol then eager protocol.

I'll replace typedef by uint16_t.

**Comment by evgeny-leksikov** [comment]
@shefty thank you for careful review! Some things will be reworked. I have a small question above regarding to coding style.

File: prov/verbs/src/ep_rdm/verbs_rdm_cm.c (hunk 5 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm_cm.c]: I looked on required sub-sequences to handle this error. I would prefer to make it in separated PR as error handling in connection management. There is a quite amount of changes not related with subject of this PR.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 1 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: > > -enum fi_ibv_rdm_buffer_status {
> > -	BUF_STATUS_FREE = 0,
> > -	BUF_STATUS_BUSY
> > -};
> > -
> > +/*
> > + * Size of the structure is performance sensitive. TODO: resize to 4
> bytes.
> > + */
> >  struct fi_ibv_rdm_buffer_service_data {
> >  	volatile enum fi_ibv_rdm_buffer_status status;
> 
> Buffer's status is a part of buffers control. Remote side is updating this
> address by RDMA_WRITE operation to release then it's can be reused.

Got it -- then my concern is that this eventually should handle endian issues.  They way I've handled endian before is to exchange he endian format as part of connection setup and do byte swapping at the initiator if the endian formats do not match.

To do this, you just need to make sure that there's a reserved flag available during connection data exchange, then the feature can be enabled later.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 7 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: > >  struct fi_ibv_rdm_header {
> > -	uint64_t imm_data;          // TODO: not implemented
> > +/*	uint64_t imm_data; TODO: not implemented */
> 
> No, this field was reserved for implementing of OFI tagged imm_data. It's
> not supported yet.

This removes a field from the structure.  Is this structure send on the wire?

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 9]
```diff
@@ -307,74 +325,51 @@ struct fi_ibv_rdm_tagged_postponed_entry {
...
-static inline enum fi_ibv_rdm_tagged_buffer_status
-fi_ibv_rdm_tagged_get_buffer_status(char *buff)
+static inline int
+fi_ibv_rdm_buffer_check_seq_num(struct fi_ibv_rdm_buf *buff, uint16_t seq_num)
 {
-	return fi_ibv_rdm_tagged_get_buff_service_data(buff)->status;
+	const int res = (seq_num == buff->service_data.seq_num);
+	VERBS_DBG(FI_LOG_EP_DATA, "seq num: %d <-> %d\n",
+		buff->service_data.seq_num, seq_num);
+	return res;
```
Comment (a-ilango) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Shouldn't returning (seq_num == buff->service_data.seq_num) be sufficient here?

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Sounds reasonable but in this case ep_rdm has more issues regarding endian format that I see because currently everything is sending with rdma write op code. CM, eager and rndv protocols also should be modified. In addition, which level should control user's data including rma and atomics? I'm guessing user?

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 7 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: yes, but this field is not handled yet properly on both sides

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 9 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: :+1:  The variable is really not needed here.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 7 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: It's changing the wire protocol.  If you update one side but not the other, the two sides will disagree on the fields that they access and break.  Before we can claim that verbs RDM support is truly supported, the wire protocol needs to be fully defined and forward/backward compatible, with proper versioning.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c (hunk 6 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: My question was why is this "err &=" versus just "err ="?  The latter is cleaner.

File: prov/verbs/src/ep_rdm/verbs_utils.h (hunk 4 above)
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_utils.h]: Note that my comment was referring to the commit overall and not just this patch.  Typically, alignment is to opening brackets/parens, though there are some exceptions where aligning to tab stops makes sense.

**Comment by shefty** [comment]
I'm fine with follow up cleanups.  Please let me know if anything needs to be addressed to this patch set before merging.

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c (hunk 6 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: OMG, indeed, here (and poll_recv) should be inverse logic with "|=".
I'll fix it with "=" and extra check in the loop statement.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 7 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: In this meaning, if I correctly understand, wire protocol should be changed soon anyway with reusing headers from fi_proto.h support shm, right? I don't think that imm support will be added earlier without explicit request.

**Comment by evgeny-leksikov** [comment]
@shefty I fixed issue with err. If you have no other objections, looks like the series is ready for merge.
